pkg/koordlet/runtimehooks/hooks/batchresource/batch_resource.go

```diff
@@ -0,0 +1,404 @@
+		p.SetContainerCFSQuota, reconciler.PodQOSFilter(), podQOSConditions...)
+	reconciler.RegisterCgroupReconciler(reconciler.ContainerLevel, sysutil.MemoryLimit, description+" (container memory limit)",
+		p.SetContainerMemoryLimit, reconciler.PodQOSFilter(), podQOSConditions...)
+}
+
+var singleton *plugin
+
+func Object() *plugin {
+	if singleton == nil {
+		singleton = &plugin{}
+	}
+	return singleton
+}
+
+func (p *plugin) SetPodResources(proto protocol.HooksProtocol) error {
+	podCtx := proto.(*protocol.PodContext)
+	if podCtx == nil {
+		return fmt.Errorf("pod protocol is nil for plugin %v", name)
+	}
+
+	err := p.SetPodCPUShares(proto)
+	if err != nil {
+		klog.V(5).Infof("failed to set pod cpu shares in plugin %s, pod %s/%s, err: %v",
+			name, podCtx.Request.PodMeta.Namespace, podCtx.Request.PodMeta.Name, err)
+	}
+
+	err1 := p.SetPodCFSQuota(proto)
+	if err1 != nil {
+		klog.V(5).Infof("failed to set pod cfs quota in plugin %s, pod %s/%s, err: %v",
+			name, podCtx.Request.PodMeta.Namespace, podCtx.Request.PodMeta.Name, err1)
+	}
+
+	err2 := p.SetPodMemoryLimit(proto)
+	if err2 != nil {
+		klog.V(5).Infof("failed to set pod memory limit in plugin %s, pod %s/%s, err: %v",
+			name, podCtx.Request.PodMeta.Namespace, podCtx.Request.PodMeta.Name, err2)
+	}
+
+	return utilerrors.NewAggregate([]error{err, err1, err2})
+}
+
+func (p *plugin) SetPodCPUShares(proto protocol.HooksProtocol) error {
+	podCtx := proto.(*protocol.PodContext)
+	if podCtx == nil {
+		return fmt.Errorf("pod protocol is nil for plugin %v", name)
+	}
+
+	if !isPodQoSBEByAttr(podCtx.Request.Labels, podCtx.Request.Annotations) {
+		return nil
+	}
+	extendedResourceSpec, err := getPodExtendedResourceSpec(podCtx.Request.Annotations)
+	if err != nil {
+		return fmt.Errorf("failed to get extendedResourceSpec, err: %v", err)
+	}
+	// if the extendedResourceSpec is empty, do nothing and keep the original cgroup configs
+	if extendedResourceSpec == nil {
+		return nil
+	}
+
+	milliCPURequest := int64(0)
+	// TODO: count init container and pod overhead
+	for _, c := range extendedResourceSpec.Containers {
+		if c.Requests == nil {
+			continue
+		}
+		q, ok := c.Requests[apiext.BatchCPU]
+		if !ok {
+			continue
+		}
+		milliCPURequest += q.Value()
+	}
+	cpuShares := milliCPURequest * sysutil.CPUShareUnitValue / 1000
+	if cpuShares < sysutil.CPUSharesMinValue {
+		cpuShares = sysutil.CPUSharesMinValue
+	}
+
+	podCtx.Response.Resources.CPUShares = pointer.Int64Ptr(cpuShares)
+	return nil
+}
+
+func (p *plugin) SetPodCFSQuota(proto protocol.HooksProtocol) error {
+	podCtx := proto.(*protocol.PodContext)
+	if podCtx == nil {
+		return fmt.Errorf("pod protocol is nil for plugin %v", name)
+	}
+
+	if !isPodQoSBEByAttr(podCtx.Request.Labels, podCtx.Request.Annotations) {
+		return nil
+	}
+	extendedResourceSpec, err := getPodExtendedResourceSpec(podCtx.Request.Annotations)
```

Comment:
if webhook is not enabled, reconciler should still work well.
how about add `extended-resource` field in protocol.PodRequest/ContainerRequest, which is inited by `FromReconciler` and `FromProxy`

Reply:
remember to support the old version protocol `recalimed-cpu/memory` in FromReconciler

Reply:
done. add ExtendedResources field and fill it if the pod/container specifies target extended resources